src/libawkward/array/IndexedArray.cpp

```diff
@@ -1022,6 +1023,41 @@ namespace awkward {
     }
   }
 
+  template <typename T, bool ISOPTION>
+  const std::shared_ptr<Content> IndexedArrayOf<T, ISOPTION>::fillna(int64_t value) const {
+    if (ISOPTION) {
+
+#if defined _MSC_VER || defined __i386__
+      std::string format = "q";
+#else
+      std::string format = "l";
+#endif
+      std::shared_ptr<int64_t> ptr(new int64_t[1], util::array_deleter<int64_t>());
+      ptr.get()[0] = value;
+      std::vector<ssize_t> shape({ 1 });
+      std::vector<ssize_t> strides({ 1 });
+      std::shared_ptr<NumpyArray> value_array = std::make_shared<NumpyArray>(Identities::none(), parameters_, ptr, shape, strides, 0, sizeof(int64_t), format);
+
+      std::vector<std::shared_ptr<Content>> contents;
+      contents.emplace_back(content());
+      contents.emplace_back(value_array);
+
+      Index8 tags = bytemask();
+      Index64 index(tags.length());
+      struct Error err = util::awkward_UnionArray_fillna_64<T>(
+        index.ptr().get(),
+        index_.ptr().get(),
+        tags.length());
+      util::handle_error(err, classname(), identities_.get());
+
+      std::shared_ptr<UnionArray8_64> out = std::make_shared<UnionArray8_64>(Identities::none(), parameters_, tags, index, contents);
+      return out;
```

Comment:
I've tried to follow your suggestion on avoiding this "check `mergeable`, then `merge` or `merge_as_union`" and run into the following: a `UnionArray` `simplify()` behaves differently when one of the contents is an `EmptyArray`. I'm checking if there is a solution for that.

Reply:
> I've tried to follow your suggestion on avoiding this "check `mergeable`, then `merge` or `merge_as_union`" and run into the following: a `UnionArray` `simplify()` behaves differently when one of the contents is an `EmptyArray`. I'm checking if there is a solution for that.

swapping the order in contents helped :-)

Reply:
Actually, if you found a bug in `UnionArray::simplify`, it would be great to fix that at the source. It's quite possible that I forgot about EmptyArray, but that's a trivial case. If an EmptyArray is in a UnionArray, it can simply be dropped, though this changes the numbering of the contents that follow and therefore the values of `tags`. The implementation of `UnionArray::simplify` rebuilds the UnionArray, renumbering `tags` as it does so. Handling EmptyArray well is a matter of checking for it and skipping it.

Reply:
No, there is no bug. The order doesn't matter. It must have been something else, I cannot reproduce the problem any more. 

It looks like the returned `Content` is the same for either "check `mergeable`, then `merge` or `merge_as_union`" or "create a `UnionArray` and return `UnionArray::simplify(true)`": its either a `NumpyArray` or a `UnionArray` of the unmergable arrays: a `NumpyArray` and a `RegularArray`, for example:
```python
<NumpyArray format="l" shape="5" data="10 10 10 10 10" at="0x7f91e4cae510"/>
<UnionArray8_64>
    <tags><Index8 i="[0 0 1]" offset="0" at="0x7f91e4cb5000"/></tags>
    <index><Index64 i="[0 1 0]" offset="0" at="0x7f91e4cae510"/></index>
    <content index="0">
        <RegularArray size="3">
            <content><NumpyArray format="d" shape="6" data="1.1 2.2 3.3 4.4 5.5 6.6" at="0x7f91e4ce9910"/></content>
        </RegularArray>
    </content>
    <content index="1">
        <NumpyArray format="l" shape="1" data="0" at="0x7f91e4c284d0"/>
    </content>
</UnionArray8_64>
<NumpyArray format="d" shape="10" data="1.1 2.2 3.3 0 0 4.4 5.5 6.6 0 0" at="0x7f91e4ceadc0"/>
<NumpyArray format="d" shape="11" data="6.9 3.9 6.9 2.2 1.5 ... 3.6 666 6.7 7.4 1.5" at="0x7f91e4cc0ca0"/>
<NumpyArray format="d" shape="12" data="777 777 1.1 777 2.2 ... 2 1 777 777 777" at="0x7f91e4d61f20"/>
```

Reply:
That's good. Creating a UnionArray and then simplifying internally checks for mergeability and then merges. The value of this is that a user might supply an integer replacement `value` to a floating point array and these types should be merged, while something truly incompatible like a number and a list should not be. All of that logic should be centralized because it's used by other operations than just this one.

Reply:
> No, there is no bug. The order doesn't matter. It must have been something else, I cannot reproduce the problem any more.

@jpivarski - I managed to trace the problem to accessing/populating a `ptr()` from C++ (the code which belongs and has been moved to `cpu-kernels`) and then using `simplify(false|true)`:

```diff
    def test_fillna_empty_array():
        empty = awkward1.layout.EmptyArray()
    
        assert awkward1.tolist(empty) == []
        array = empty.rpad(5, 0)
        assert awkward1.tolist(array) == [None, None, None, None, None]
>       assert awkward1.tolist(array.fillna(10)) == [10, 10, 10, 10, 10]
E       assert [43118103050, 168430090, 657930, 2570, 10] == [10, 10, 10, 10, 10]
E         At index 0 diff: 43118103050 != 10
E         Full diff:
E         - [43118103050, 168430090, 657930, 2570, 10]
E         + [10, 10, 10, 10, 10]

tests/test_0072_fillna_operation.py:16: AssertionError
```